frontend/model/contracts/chatroom.js

```diff
@@ -0,0 +1,363 @@
+}
+
+const chatRoomType: any = objectOf({
+  attributes: objectOf(chatRoomAttributes)
+})
+
+export const messageType: any = objectMaybeOf({
+  type: unionOf(...Object.values(MESSAGE_TYPES).map(v => literalOf(v))),
+  text: string, // message text | proposalId when type is INTERACTIVE | notificationType when type if NOTIFICATION
+  notification: objectMaybeOf({
+    type: unionOf(...Object.values(MESSAGE_NOTIFICATIONS).map(v => literalOf(v))),
+    params: mapOf(string, string) // { username }
+  }),
+  replyingMessage: objectOf({
+    id: string, // scroll to the original message and highlight
+    index: number, // index of the list of messages
+    username: string, // display username
+    text: string // display text(if too long, truncate)
+  }),
+  emoticons: mapOf(string, arrayOf(string)), // mapping of emoticons and usernames
+  onlyVisibleTo: arrayOf(string) // list of usernames, only necessary when type is NOTIFICATION
+  // TODO: need to consider POLL and add more down here
+})
+
+export function createMessage ({ meta, data, hash, state }: {
+  meta: Object, data: Object, hash: string, state?: Object
+}): Object {
+  const { type, text, replyingMessage } = data
+  const { createdDate } = meta
+
+  let newMessage = {
+    type,
+    time: new Date(createdDate).toISOString(),
+    id: hash,
+    from: meta.username
+  }
+
+  if (type === MESSAGE_TYPES.TEXT) {
+    newMessage = !replyingMessage ? { ...newMessage, text } : { ...newMessage, text, replyingMessage }
+  } else if (type === MESSAGE_TYPES.POLL) {
+    // TODO: Poll message creation
+  } else if (type === MESSAGE_TYPES.NOTIFICATION) {
+    const params = {
+      channelName: state?.attributes.name,
+      channelDescription: state?.attributes.description,
+      ...data.notification
+    }
+    delete params.type
+    newMessage = {
+      ...newMessage,
+      notification: { type: data.notification.type, params }
+    }
+  } else if (type === MESSAGE_TYPES.INTERACTIVE) {
+    // TODO: Interactive message creation for proposals
+  }
+  return newMessage
+}
+
+export function getLatestMessages ({
+  count, messages
+}: { count: number, messages: Array<Object> }): Array<Object> {
+  return messages.slice(Math.max(messages.length - count, 0))
+}
+
+export async function leaveChatRoom ({ contractID }: {
+  contractID: string
+}) {
+  const rootState = sbp('state/vuex/state')
+  const rootGetters = sbp('state/vuex/getters')
+  if (contractID === rootGetters.currentChatRoomId) {
+    await sbp('state/vuex/commit', 'setCurrentChatRoomId', {
+      groupId: rootState.currentGroupId
+    })
+    const curRouteName = sbp('controller/router').history.current.name
+    if (curRouteName === 'GroupChat' || curRouteName === 'GroupChatConversation') {
+      sbp('controller/router')
+        .push({ name: 'GroupChatConversation', params: { chatRoomId: rootGetters.currentChatRoomId } })
+        .catch(logExceptNavigationDuplicated)
```

Comment:
Actually, I am not sure what would be broken later because of this missing `await` here. But added anyway just to prevent for the unexpected condition.